sdk/storage/azure-storage-blob-nio/src/test/java/com/azure/storage/blob/nio/AzureFileSystemProviderSpec.groovy

```diff
@@ -303,15 +306,509 @@ class AzureFileSystemProviderSpec extends APISpec {
+
+        then:
+        thrown(DirectoryNotEmptyException)
+        ((AzureFileSystemProvider) fs.provider()).checkDirectoryExists(destinationClient)
+
+        where:
+        destinationIsVirtual | _
+        true                 | _
+        false                | _
+    }
+
+    @Unroll
+    def "FileSystemProvider copy replace existing fail"() {
+        // The success case is tested by the "copy destination" test.
+        // Testing replacing a virtual directory is in the "non empty dest" test as there can be no empty virtual dir.
+        setup:
+        def fs = createFS(config)
+
+        // Generate resource names.
+        // Don't use default directory to ensure we honor the root.
+        def rootName = fs.getRootDirectories().last().toString()
+        def containerName = rootToContainer(rootName)
+        def sourceName = generateBlobName()
+        def sourcePath = fs.getPath(rootName, sourceName)
+        def destName = generateBlobName()
+        def destPath = fs.getPath(rootName, destName)
+
+        // Generate clients to resources.
+        def containerClient = primaryBlobServiceClient
+            .getBlobContainerClient(containerName)
+        def sourceClient = containerClient.getBlobClient(sourceName)
+        def destinationClient = containerClient.getBlobClient(destName)
+
+        // Create resources as necessary
+        sourceClient.upload(new ByteArrayInputStream(getRandomByteArray(20)), 20)
+        if (destinationIsDir) {
+            fs.provider().createDirectory(destPath)
+        } else {
+            destinationClient.upload(defaultInputStream.get(), defaultDataSize)
+        }
+
+        when:
+        fs.provider().copy(sourcePath, destPath, StandardCopyOption.COPY_ATTRIBUTES)
+
+        then:
+        thrown(FileAlreadyExistsException)
+        if (destinationIsDir) {
+            assert ((AzureFileSystemProvider) fs.provider()).checkDirectoryExists(destinationClient)
+        } else {
+            def outStream = new ByteArrayOutputStream()
+            destinationClient.download(outStream)
+            assert ByteBuffer.wrap(outStream.toByteArray()) == defaultData
+        }
+
+        where:
+        destinationIsDir | _
+        true             | _
+        false            | _
+        // No need to test virtual directories. If they exist, they aren't empty and can't be overwritten anyway.
+        // See above.
+    }
+
+    def "FileSystemProvider copy options fail"() {
+        setup:
+        def fs = createFS(config)
+
+        // Generate resource names.
+        // Don't use default directory to ensure we honor the root.
+        def rootName = fs.getRootDirectories().last().toString()
+        def sourceName = generateBlobName()
+        def sourcePath = fs.getPath(rootName, sourceName)
+        def destName = generateBlobName()
+        def destPath = fs.getPath(rootName, destName)
```

Comment:
why dont we add this repeated to the setup and create global variables?

Reply:
I guess its not really repeated in every test

Reply:
I didn't quite follow this suggestion because I think the tests in this spec will vary too much to make them global, but I did a lot of refactoring of tests and cut out a lot of lines. That'll show up in my next PR.